### PR TITLE
STR-544: Add new popover assets  to RSUIKit

### PR DESCRIPTION
**Pull request by calebabm**
Add Revel Button, Popover components, checkmark and X sf symbols 

Notes: 
Built with SwiftUI 
Does not contain logic for branding tool customization. 
Screen Shot 2022-02-23 at 11 28 36 AM

Screen Shot 2022-02-23 at 11 22 56 AM

### CODE REVIEW

File: Sources/RSUIKit/Supporting/K/ImagesConstants.swift [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+//  ImagesConstants.swift
+//  
+//
+//  Created by Caleb Meurer on 2/17/22.
+//
+
+import Foundation
+import SwiftUI
+
+public enum ImagesConstants {
```
Comment (paulvaldes) [inline Sources/RSUIKit/Supporting/K/ImagesConstants.swift:11]: Did you mean for `Images` to be plural ?

File: Sources/RSUIKit/Supporting/K/ImagesConstants.swift [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        case xCircleFill(color: Color?)
+        
+        
+        public var image: AnyView {
+            switch self {
+            case .checkmarkCircleFill(let color):
+                return AnyView(Image(systemName: "checkmark.circle.fill")
+                                .resizable()
+                                .foregroundColor(color)
+                                .frame(width: 48, height: 48, alignment: .center))
```
Comment (paulvaldes) [inline Sources/RSUIKit/Supporting/K/ImagesConstants.swift:23]: Are these size values going to be used elsewhere or ever changed? If not, possibly use a `private let iconHeight = 48`?

File: Sources/RSUIKit/Supporting/K/SizeConstants.swift [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+enum SizeConstants {
+    
+    enum Button<T: View> {
+        case alert(view: T)
+        
+        var frame: some View {
+            switch self {
+            case .alert(let view):
+                return view.frame(width: 105, height: 39)
```
Comment (paulvaldes) [inline Sources/RSUIKit/Supporting/K/SizeConstants.swift:18]: Are these values static regardless of device, including iPhone vs iPad?

File: Sources/RSUIKit/Alerts/PopoverView.swift [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            }
+            renderDetails()
+        }
+        SizeConstants.Alert.popover(view: view).frame
+        .background(Color.white)
+        .clipShape(RoundedRectangle(cornerRadius: 15))
+    }
+    
+    private func constructButtons() -> some View {
+        var buttons: [RSUIButton] = []
```
Comment (paulvaldes) [inline Sources/RSUIKit/Alerts/PopoverView.swift:29]: Are we going to handle the number of buttons in a certain way; e.g. the user adds 7 buttons > what happens > show the first 3? show the last 3? show all 7? Or would this just be misuse of the API and not included in the framework API structure/limitations?

**Review by paulvaldes — APPROVED** [review]
Looks good, crazy how few lines of code this is

File: Sources/RSUIKit/Buttons/RSUIButton.swift [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+//
+
+import SwiftUI
+
+@available(iOS 14.0, *)
+public struct RSUIButton: View {
+    public var title: String
+    private var action: () -> Void
+    @State private var tapped = false
+
```
[the remark this answers is not in the document]
Comment (calebabm) [reply Sources/RSUIKit/Buttons/RSUIButton.swift:15]: Basically all views in SwiftUI are identifiable, so there is no need to use a tag 
check this out for more info: [WWDC21](https://developer.apple.com/videos/play/wwdc2021/10022)

File: Sources/RSUIKit/Buttons/RSUIButton.swift [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        let button = Button {
+            tapped.toggle()
+            action()
+        } label: {
+            Text(title)
+                .foregroundColor(.revelBlue)
+        }
+        SizeConstants.Button.alert(view: button).frame
+        .overlay(
+            RoundedRectangle(cornerRadius: 5)
```
[the remark this answers is not in the document]
Comment (calebabm) [reply Sources/RSUIKit/Buttons/RSUIButton.swift:31]: Probably not, as this is just how rounded the element is, which seems to be static

File: Sources/RSUIKit/Supporting/K/ImagesConstants.swift (hunk 2 above)
Comment (calebabm) [reply Sources/RSUIKit/Supporting/K/ImagesConstants.swift:23]: I was thinking we may want to pass in the size, when we want to use. So we would have color, width, height, and alignment

File: Sources/RSUIKit/Supporting/K/SizeConstants.swift (hunk 3 above)
Comment (calebabm) [reply Sources/RSUIKit/Supporting/K/SizeConstants.swift:18]: SwiftUI optimizes based on what device, so it should general look the same no matter what, however this should be verified with UX just to make sure it's what they want

File: Sources/RSUIKit/Alerts/PopoverView.swift (hunk 4 above)
Comment (calebabm) [reply Sources/RSUIKit/Alerts/PopoverView.swift:29]: Basically abusing the API, and pretty sure apple's User Interface Guidelines. I will bring it up in the architecture meetings though

**Comment by dkulemalin** [comment]
Just general question. Do you have UI Design for your feature? Popover on screenshot differs from what we have on POS right now (paddings between buttons, background color, text colors are wrong). So I do not think we can reuse it on POS. or it is some draft version of popover?

File: Sources/RSUIKit/Buttons/RSUIButton.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (paulvaldes) [reply Sources/RSUIKit/Buttons/RSUIButton.swift:15]: I am more so talking about the ability to store a value in the button object that can be used to access, modify, configure, etc the given button. 

For example:
In `ScalesService` we have a set of tags defined which are then used to do various things
```
#define TagMessageViewManualWeight 101
#define TagMessageViewTare 102
#define TagMessageViewScales 103
#define TagMessageViewConnectionStatus 104
#define TagMessageViewInputTareWeight 105 

typedef enum {
    TagCancelButton = 0,
    TagManualWeightButton = 1,
    TagOkButton = 2,
    TagMultiplyTareButton = 3,
    TagSetTareButton = 4,
    TagInputTareWeight = 5,
} ScaleServiceButtonTags;

```

File: Sources/RSUIKit/Buttons/RSUIButton.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (paulvaldes) [reply Sources/RSUIKit/Buttons/RSUIButton.swift:31]: Sorry, I meant has UI given a set of corner radius' that are to be used across various elements or is it specific to each element

**Comment by calebabm** [comment]
> Just general question. Do you have UI Design for your feature? Popover on screenshot differs from what we have on POS right now (paddings between buttons, background color, text colors are wrong). So I do not think we can reuse it on POS. or it is some draft version of popover?

Yep. here is the figma link: [https://www.figma.com/file/ignDosOLmLN8aj3nbgjxPo/STR-168?node-id=114%3A18506 - Connect to preview](https://www.figma.com/file/ignDosOLmLN8aj3nbgjxPo/STR-168?node-id=114%3A18506)

**Comment by dkulemalin** [comment]
@calebabm 
So in figma it looks better.
The question to designer can be: why popover on design differs from all other popovers we use in the app. buttons are aligned by center, while in the rest app we do differently. And no header.

But anyway your screenshot differs from mockups in figma

**Comment by calebabm** [comment]
> @calebabm So in figma it looks better. The question to designer can be: why popover on design differs from all other popovers we use in the app. buttons are aligned by center, while in the rest app we do differently. And no header.
> 
> But anyway your screenshot differs from mockups in figma

Oh yeah, i see what your saying. Looks like the background color and title weight. Will update and attach screen shots

**Comment by calebabm** [comment]
Update to match the original ticket's design [STR-428](https://revelup.atlassian.net/browse/STR-428)

File: Sources/RSUIKit/Alerts/PopoverView.swift [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        .frame(width: SizeConstants.Alert.popover.frame.width, height: SizeConstants.Alert.popover.frame.height)
+        .background(Color.revelBlueLt)
+        .clipShape(RoundedRectangle(cornerRadius: 10))
+    }
+    
+    @ViewBuilder
+    var renderDetails: some View {
+        VStack {
+            Text(dataSource.title)
+                .fontWeight(.medium)
```
Comment (dkulemalin) [inline Sources/RSUIKit/Alerts/PopoverView.swift:31]: in design title and subtitle have astronautBlue color

File: Sources/RSUIKit/Buttons/RSUIButton.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (paulvaldes) [reply Sources/RSUIKit/Buttons/RSUIButton.swift:15]: What was the resolution on all of this?

File: Sources/RSUIKit/Alerts/PopoverView.swift [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        VStack {
+            if (dataSource.imageView != nil) {
+                dataSource.imageView
+                    .padding([.top, .bottom], 20)
+            }
+            renderDetails
+        }
+        .frame(width: SizeConstants.Alert.popover.frame.width, height: SizeConstants.Alert.popover.frame.height)
+        .background(Color.revelBlueLt)
+        .clipShape(RoundedRectangle(cornerRadius: 10))
```
Comment (danikkm) [inline Sources/RSUIKit/Alerts/PopoverView.swift:24]:
```swift
.clipShape(RoundedRectangle(cornerRadius: 10, style: .continuous))
```
iOS uses a continuous curving style. If you are interested, take a look at the following article: https://99percentinvisible.org/article/circling-square-designing-squircles-instead-rounded-rectangles. 
What do you think, should we use `style: .continuous` as well? If we decide to opt-in, apply changes to every rounded shape.

File: Sources/RSUIKit/Models/PopoverDataSource.swift [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+//  
+//
+//  Created by Caleb Meurer on 2/17/22.
+//
+
+import SwiftUI
+
+public struct PopoverDataSource {
+    
+    var imageView: AnyView?
```
Comment (danikkm) [inline Sources/RSUIKit/Models/PopoverDataSource.swift:12]: Is there a need to use AnyView? You can create a specific type of view here. Naming implies that it stores an image.

File: Sources/RSUIKit/Alerts/PopoverView.swift (hunk 7 above)
Comment (calebabm) [reply Sources/RSUIKit/Alerts/PopoverView.swift:31]: ahh yep, thank you!

File: Sources/RSUIKit/Alerts/PopoverView.swift (hunk 8 above)
Comment (calebabm) [reply Sources/RSUIKit/Alerts/PopoverView.swift:24]: ohh yeah! that's a great idea, thank you!

File: Sources/RSUIKit/Models/PopoverDataSource.swift (hunk 9 above)
Comment (calebabm) [reply Sources/RSUIKit/Models/PopoverDataSource.swift:12]: Ahh yeah, will update the naming. In this situation, with the image constants configuring the attributes of the image returning `some View`, wrapping in `AnyView` made it more readable from the call site of `PopoverDataSource`

File: Sources/RSUIKit/Alerts/PopoverView.swift [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+//  
+//
+//  Created by Caleb Meurer on 2/17/22.
+//
+
+import UIKit
+import SwiftUI
+
+@available(iOS 14.0, *)
+public struct PopoverView: View {
```
Comment (dkulemalin) [inline Sources/RSUIKit/Alerts/PopoverView.swift:12]: Should we name it RSUIPopoverView for consistency?

File: Sources/RSUIKit/Models/PopoverDataSource.swift (hunk 9 above)
Comment (danikkm) [reply Sources/RSUIKit/Models/PopoverDataSource.swift]: AnyView removes identity from the view, making this view not unique in a view hierarchy.

**Review by danikkm — APPROVED** [review]

**Pull request merged**